File: typescipt/existing/src/language/expressions/functions/newFunction.ts






namespace Lexy.Compiler.Language.Expressions.Functions;

public class NewFunction : ExpressionFunction, IHasNodeDependencies
{
   public const string Name = "new";

   protected string FunctionHelp => $"{Name} expects 1 argument (Function.Parameters)";

   public MemberAccessLiteral TypeLiteral { get; }

   public Expression ValueExpression { get; }

   public ComplexTypeReference Type { get; private set; }

   private NewFunction(Expression valueExpression, SourceReference reference)
     : base(reference)
   {
     ValueExpression = valueExpression ?? throw new ArgumentNullException(nameof(valueExpression));
     TypeLiteral = (valueExpression as MemberAccessExpression)?.MemberAccessLiteral;
   }

   public IEnumerable<IRootNode> GetDependencies(RootNodeList rootNodeList)
   {
     if (Type ! null) yield return rootNodeList.GetNode(Type.Name);
   }

   public static ExpressionFunction Create(SourceReference reference, Expression expression)
   {
     return new NewFunction(expression, reference);
   }

   public override IEnumerable<INode> GetChildren()
   {
     yield return ValueExpression;
   }

   protected override void Validate(IValidationContext context)
   {
     var valueType = ValueExpression.DeriveType(context);
     if (!(valueType is ComplexTypeReference complexTypeReference))
     {
       context.Logger.Fail(Reference,
         $"Invalid argument 1 'Value' should be of type 'ComplexTypeType' but is 'ValueType'. {FunctionHelp}");
       return;
     }

     Type = complexTypeReference;
   }

   public override VariableType DeriveReturnType(IValidationContext context)
   {
     var nodeType = context.RootNodes.GetType(TypeLiteral.Parent);
     var typeReference = nodeType?.MemberType(TypeLiteral.Member, context) as ComplexTypeReference;
     return typeReference?.GetComplexType(context);
   }
}

File: src/Lexy.Compiler/Language/Expressions/Functions/NewFunction.cs
using System;
using System.Collections.Generic;
using Lexy.Compiler.Language.VariableTypes;
using Lexy.Compiler.Parser;
using Lexy.Compiler.Parser.Tokens;

namespace Lexy.Compiler.Language.Expressions.Functions;

public class NewFunction : FunctionCallExpression, IHasNodeDependencies
{
    public const string Name = "new";

    protected string FunctionHelp => $"{Name} expects 1 argument new(Function.Parameters)";

    public MemberAccessLiteral TypeLiteral { get; }

    public Expression ValueExpression { get; }

    public ComplexType Type { get; private set; }

    private NewFunction(Expression valueExpression, ExpressionSource source)
        : base(Name, source)
    {
        ValueExpression = valueExpression ?? throw new ArgumentNullException(nameof(valueExpression));
        TypeLiteral = (valueExpression as MemberAccessExpression)?.MemberAccessLiteral;
    }

    public IEnumerable<IRootNode> GetDependencies(IRootNodeList rootNodeList)
    {
        if (Type != null) yield return rootNodeList.GetNode(Type.Name);
    }

    public static FunctionCallExpression Create(ExpressionSource source, Expression expression)
    {
        return new NewFunction(expression, source);
    }

    public override IEnumerable<INode> GetChildren()
    {
        yield return ValueExpression;
    }

    protected override void Validate(IValidationContext context)
    {
        var valueType = ValueExpression.DeriveType(context);
        if (valueType is not ComplexType complexType)
        {
            context.Logger.Fail(Reference,
                $"Invalid argument 1. 'Value' should be of type 'ComplexType' but is '{valueType?.GetType()}'. {FunctionHelp}");
            return;
        }

        Type = complexType;
    }

    public override VariableType DeriveType(IValidationContext context)
    {
        var nodeType = context.RootNodes.GetType(TypeLiteral.Parent);
        return nodeType?.MemberType(TypeLiteral.Member, context.RootNodes) as ComplexType;
    }
}